streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamPartitionAssignorTest.java

```
@@ -895,25 +891,30 @@ public Object apply(Object value1, Object value2) {
+        final MockClientSupplier mockClientSupplier = new MockClientSupplier();
+        final StreamThread streamThread = new StreamThread(builder, config, mockClientSupplier, applicationId, client, uuid, new Metrics(), Time.SYSTEM, new StreamsMetadataState(builder, StreamsMetadataState.UNKNOWN_HOST), 0);
 
         final StreamPartitionAssignor partitionAssignor = new StreamPartitionAssignor();
         partitionAssignor.configure(config.getConsumerConfigs(streamThread, applicationId, client));
+        final MockInternalTopicManager mockInternalTopicManager = new MockInternalTopicManager(streamThread.config, mockClientSupplier.restoreConsumer);
+        partitionAssignor.setInternalTopicManager(mockInternalTopicManager);
 
         final Map<String, PartitionAssignor.Subscription> subscriptions = new HashMap<>();
-        final Set<TaskId> emptyTasks = Collections.<TaskId>emptySet();
+        final Set<TaskId> emptyTasks = Collections.emptySet();
         subscriptions.put(
             client,
             new PartitionAssignor.Subscription(
                 Collections.singletonList("unknownTopic"),
                 new SubscriptionInfo(uuid, emptyTasks, emptyTasks, userEndPoint).encode()
             )
         );
-        // TODO: Update the code accordingly,
-        // This line was added to fix the test failure since internalTopicManager is created in the config method all the time.
-        partitionAssignor.internalTopicManager = null;
         final Map<String, PartitionAssignor.Assignment> assignment = partitionAssignor.assign(metadata, subscriptions);
 
+        final Map<String, Integer> expectedCreatedInternalTopics = new HashMap<>();
+        expectedCreatedInternalTopics.put(applicationId + "-KSTREAM-JOINOTHER-0000000012-store-changelog", 3);
```

Thinking about this once more, I think, there is still an issue. As we do not know the number the input topic partitions for one join input, we cannot know the number of task for the join, and thus, can actually not create this two changelog topics either.

\cc @guozhangwang WDYT?

---

If that was the case wouldn't we get an exception due to the topics not being co-partitioned? We should have a test for this case.

---

When some expected external topics do not exist or do not have metadata updatable, current strategy is that Streams will not create any tasks of its "affected" sub-topologies, including the changelog topic and repartition topic as well, and will expect another rebalance to trigger once these topics shows up. If the source topics are input ones for the "head sub-topologies" then no tasks will be created and no internal topics will be created.

---

As I expected. But this strategy does not work correctly...The tasks for join-subtopology should not get created (an neither this two changelog topics). Will investigate an update this PR.

---

@dguy After metadata refresh, we might end up with a problem. With the latest changes, we should throw an exception as number of partitions would not match the expected number. I am not sure if we can have an extra test for this -- from my understanding, this test should cover the case (and maybe other). I am not sure, if we cover all cases...